content/post/fewer-parameters-means-fewer-bugs.md

```diff
@@ -0,0 +1,244 @@
+=> "Ruby is the best - Chen, Bater (2020)"
+  
+robert = Author.new("Robert", "Kiyosaki", "T")
+Book.new("Rich Dad, Poor Dad", "1997", robert).summary
+=> "Rich Dad, Poor Dad - Kiyosaki, T, Robert (1997)"
+```
+In this design, the existing behavior doesn't change. The default value of the middle name is `nil` (also known as Null in other languages) because in most cases we can skip it and we won't break anything when the existing code tries to create a new author with only two parameters. The author class is open for extension and the book class is closed for modification in this case. Without refactoring, we need to touch upon both the `summary` and `cover` methods, causing the parameters to increase from two to three. Let's compare the code before and after, again:
+#### Old pattern without refactoring
+```rb
+def summary(first_name, last_name, middle_name = nil)
+  if middle_name.nil?
+    "#{title} - #{last_name}, #{first_name} (#{year_published})"
+  else
+    "#{title} - #{last_name}, #{middle_name}, #{first_name} (#{year_published})"
+  end
+end
+```
+#### After refactoring
+```rb
+def summary
+  "#{title} - #{author.full_name} (#{year_published})"
+end
+```
+Obviously, the code quality has a huge improvement after refactoring.
+
+## Loose coupling
+Before we finish refactoring and send off this PR to someone, let's check the code again. But is it clean enough?
+
+The code seems short and simple, but this method still depends on the author class. In other words, the `summary` method needs to know the author object has a `full_name` method, we could say the book class still couple with author class. In this case, maybe it's hard to notice the cost of coupling, let's imagine a few years later, the `summary` method might become more complicated:
+
+#### Few years later
+```rb
+def summary
+  # 10 lines of code you don't want to read.
+  middle_result = "#{some_author_title_helper(author)} - #{author.full_name}"
+  # another 10 lines of code
+  return final_result
+end
+```
+The disadvantage of coupling is that when one class needs to change, you need to change others as well. For example, one day we need to rename the full name method or add more parameters to it, we need to modify the book class at the same time, which we don't like.
+
+The bad news is, in the real world it's almost impossible to remove all the coupling between each object, but we can at least reduce the impact by segregating and highlighting the dependence. 
+
+```rb
+def summary
+  "#{title} - #{author_name} (#{year_published})"
+end
+
+def author_name
+  author.full_name
+end
+```
+After this refactoring, the `summary` method 100% calling from the book object itself. The book class still has the coupling with the author class, but it's obvious and easy to maintain. If we only consider the example case, you might wonder is that a little bit over design? Well, I would say that a good pattern is never too early. Now we can push the PR confidently and enjoy the holiday.
```

Comment:
I would highlight the trade-off more explicitly. We get more explicit coupling (benefit) but we have to accept indirection (drawback). Indirection makes it harder to read & understand code (just like long lists of parameters).